File: optimal/72.R
# Counting fractions

options(scipen=100)

Args  <- commandArgs()
limit <- ifelse(is.na(Args[6]), 1000000, as.numeric(Args[6]))

sieve_limit <- (floor(sqrt(limit)) - 1) %/% 2
if ( sieve_limit == 0 ) {
    sieve_limit <- 1
}
max_index <- (limit - 1) %/% 2
cache <- rep(0, max_index)
for ( n in seq.int(sieve_limit) ) {
    if ( cache[n] == 0 ) {
        k <- 2 * n * (n + 1)
        p <- 2 * n + 1
        while ( k <= max_index ) {
            if ( cache[k] == 0 ) {
                cache[k] <- p
            }
            k <- k + p
        }
    }
}
multiplier <- 1
while ( multiplier <= limit ) {
    multiplier <- multiplier * 2
}
multiplier <- multiplier %/% 2
count <- multiplier - 1
multiplier <- multiplier %/% 2
step_index <- ((limit %/% multiplier) + 1) %/% 2
for ( n in seq.int(max_index) ) {
    if ( n == step_index ) {
        multiplier <- multiplier %/% 2
        step_index <- ((limit %/% multiplier) + 1) %/% 2
    }
    if ( cache[n] == 0 ) {
        cache[n] <- 2 * n
        count <- count + multiplier * cache[n]
    } else {
        p <- cache[n]
        cofactor <- (2 * n + 1) %/% p
        factor <- p
        if ( cofactor %% p ) {
            factor <- p - 1
        }
        cache[n] <- factor * cache[cofactor %/% 2]
        count <- count + multiplier * cache[n]
    }
}

cat(count, fill=TRUE)

File: optimal/72.js
#!/usr/bin/env node

// Counting fractions

var program = require('commander');

program
    .version('0.1.0')
    .description('Counting fractions')
    .option('-l, --limit <int>', 'The limit on d', Number, 1000000)
    .parse(process.argv);

var sieve_limit = Math.floor(Math.floor(Math.sqrt(program.limit - 1)) / 2);
var max_index = Math.floor((program.limit - 1) / 2);
var cache = [];
for (var n = 1; n <= sieve_limit; n++) {
    if (!cache[n]) {
        var p = 2 * n + 1;
        var k = 2 * n * (n + 1);
        while (k <= max_index) {
            if (!cache[k]) {
                cache[k] = p;
            }
            k += p;
        }
    }
}
var multiplier = 1;
while (multiplier <= program.limit) {
    multiplier *= 2;
}
multiplier /= 2;
var count = multiplier - 1;
multiplier /= 2;
var step_index = Math.floor((Math.floor(program.limit / multiplier) + 1) / 2);
for (var n = 1; n <= max_index; n++) {
    if (n == step_index) {
        multiplier /= 2;
        step_index =
            Math.floor((Math.floor(program.limit / multiplier) + 1) / 2);
    }
    if (!cache[n]) {
        cache[n] = 2 * n;
        count += multiplier * cache[n];
    } else {
        var p = cache[n];
        var cofactor = Math.floor((2 * n + 1) / p);
        var factor = p;
        if (cofactor % p) {
            factor = p - 1;
        }
        cache[n] = factor * cache[Math.floor(cofactor / 2)];
        count += multiplier * cache[n];
    }
}

console.log(count);
